snuba/query/matcher.py

```diff
@@ -0,0 +1,218 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
+from datetime import date, datetime
+from typing import Generic, Mapping, Optional, Sequence, Tuple, TypeVar
+
+from snuba.query.expressions import Column as ColumnExpr
+from snuba.query.expressions import Expression, FunctionCall
+
+TNode = TypeVar("TNode")
+
+
+@dataclass(frozen=True)
+class MatchResult:
+    """
+    Contains the parameters found in an expression by a Pattern.
+    Parameters are the equivalent to groups in regular expressions, they
+    are named parts of the expression we are matching that are identified
+    when a valid match is found.
+    """
+
+    expressions: Mapping[str, Expression] = field(default_factory=dict)
+    strings: Mapping[str, Optional[str]] = field(default_factory=dict)
+    ints: Mapping[str, Optional[int]] = field(default_factory=dict)
+    floats: Mapping[str, Optional[float]] = field(default_factory=dict)
+    dates: Mapping[str, Optional[date]] = field(default_factory=dict)
+    datetimes: Mapping[str, Optional[datetime]] = field(default_factory=dict)
+
+    def merge(self, values: MatchResult) -> MatchResult:
+        return MatchResult(
+            expressions={**self.expressions, **values.expressions},
+            strings={**self.strings, **values.strings},
+            ints={**self.ints, **values.ints},
+            floats={**self.floats, **values.floats},
+            dates={**self.dates, **values.dates},
+            datetimes={**self.datetimes, **values.datetimes},
```

Comment:
Is there a reason not to use a `ChainMap` here?

Reply:
None, since these are generally tiny and mostly empty dicts I went for readability (I find this more readable now that I know what it means) over performance. I can move to Chainmap if you like it more.

Reply:
Not a big deal, I think if the dictionary is collapsed into a single one it's less of a problem. It might be helpful to have the chain map around if we end up having to debug any parameter name shadowing issues but otherwise I don't think there's a significant enough benefit to justify choosing one over the other.